### PR TITLE
fix: onbjerg's typed ast integration

### PR DESCRIPTION
**Pull request by iFrostizz**
## Motivation

Following up on https://github.com/gakonst/ethers-rs/pull/1943 
This caused lots of issues, shown in this PR: #1966 
And modifying in order to integrate it in foundry without breaking compatibility: https://github.com/foundry-rs/foundry/pull/3989

## Solution

Recover some typings of the ast into the old ones, and providing an helper function to convert from old to typed ast.
This is still very much a WIP but open for thoughts, mainly for the old => typed ast function `to_typed()`.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -20,9 +20,9 @@ checksum = "f26201604c87b1e01bd3d98f8d5d9a8fcbb815e8cedb41ffccbeb4bf593a35fe"
 
 [[package]]
```
Comment (gakonst) [inline Cargo.lock:21]: unrelated changes in deps should not be part of the pr

Comment (iFrostizz) [reply Cargo.lock:21]: Not sure what happened here, the github view also is broken there might be lost commit for some reason and it updated packages in the middle of it. Else I have conflicts in the `Cargo.lock`

**Comment by gakonst** [comment]
hey @iFrostizz thanks for giving this another shot. there seem to be a bunch of merge conflicts which I'm not sure how you got. 

can we please break this in smaller PRs that we can individually review w/o merge conflicts? this is currently too big / requires too much context to get it reasonably over the line.

**Review by mattsse — CHANGES_REQUESTED** [review]
please rebase master, should get rid of all changes that are already on master

**Comment by iFrostizz** [comment]
Sure will do that @gakonst ! Also, I could work on the ast visitor, had the idea to write helpers with the ability to filter node types.
@mattsse , I believe that I did rebased from master. Then cherry-picked the already commited ones. Did I chose a too old commit ?

**Comment by mattsse** [comment]
this branch has a couple of conflicts that must be resolved after rebase|merge master.

**Comment by iFrostizz** [comment]
seems like the merge did it this time

File: ethers-solc/src/artifacts/ast/lowfidelity.rs [hunk 2]
```diff
@@ -24,6 +24,14 @@ pub struct Ast {
     pub other: BTreeMap<String, serde_json::Value>,
 }
 
+impl Ast {
+    // TODO: find a better name
+    pub fn to_typed(self) -> SourceUnit {
+        let data = serde_json::to_string(&self).unwrap();
+        serde_json::from_str(&data).unwrap()
+    }
```
Comment (gakonst) [inline ethers-solc/src/artifacts/ast/lowfidelity.rs:32]: @mattsse shouldn't we be converting manually from the internal types, instead of ser/de'ing? or am i missing something?

Comment (iFrostizz) [reply ethers-solc/src/artifacts/ast/lowfidelity.rs:32]: That would probably imply to write (or reuse, maybe in coverage) a node walker to convert (de-reserialize) every node so not sure if the performance is too bad doing that ?

**Comment by iFrostizz** [comment]
Friends @mattsse , @gakonst , can we consider to pull this PR ?
Also, I have made some notable changes in a new branch that provides fixes for the ast as well as a visitor trait to traverse the AST easily. Should I open a new one ?

**Comment by tynes** [comment]
I like the typed ast integration because it will make it easier to do `forge inspect immutableReferences` as the solc output doesn't include the immutable references for contracts that are inherited from. You need to traverse the ast to find these contracts and accumulate all of the immutable references

**Comment by iFrostizz** [comment]
> I like the typed ast integration because it will make it easier to do `forge inspect immutableReferences` as the solc output doesn't include the immutable references for contracts that are inherited from. You need to traverse the ast to find these contracts and accumulate all of the immutable references

Wrote a new branch [`franfran/visitor`](https://github.com/iFrostizz/ethers-rs/tree/franfran/visitor) which includes a visitor that can be implemented to ease the writing of yet another one (there's already one in forge fmt).
https://github.com/iFrostizz/ethers-rs/blob/franfran/visitor/ethers-solc/src/artifacts/ast/visitor.rs

**Comment by gakonst** [comment]
Closing as stale. Please open a new PR against the latest branch

**Comment by wtdcode** [comment]
> > I like the typed ast integration because it will make it easier to do `forge inspect immutableReferences` as the solc output doesn't include the immutable references for contracts that are inherited from. You need to traverse the ast to find these contracts and accumulate all of the immutable references
> 
> Wrote a new branch [`franfran/visitor`](https://github.com/iFrostizz/ethers-rs/tree/franfran/visitor) which includes a visitor that can be implemented to ease the writing of yet another one (there's already one in forge fmt). https://github.com/iFrostizz/ethers-rs/blob/franfran/visitor/ethers-solc/src/artifacts/ast/visitor.rs

I tried your branch but it seems lacking `end_visit` functions.
